theseus/embodied/robot/robot.py

```diff
@@ -0,0 +1,175 @@
+            elif joint.type == "prismatic":
+                return PrismaticJoint
+            elif joint.type == "fixed":
+                return FixedJoint
+            else:
+                raise ValueError(f"{joint.type} is currently not supported.")
+
+        def get_origin(urdf_origin: Optional[urdf.Pose] = None) -> torch.Tensor:
+            origin = torch.eye(3, 4, dtype=dtype).unsqueeze(0)
+            if urdf_origin is None:
+                return origin
+
+            if urdf_origin.xyz is not None:
+                origin[:, :, 3] = torch.tensor(urdf_origin.xyz, dtype=dtype)
+
+            if urdf_origin.rpy is not None:
+                rpy = urdf_origin.rpy
+                rot_x = so3.exp(torch.tensor([[rpy[0], 0, 0]], dtype=dtype))
+                rot_y = so3.exp(torch.tensor([[0, rpy[1], 0]], dtype=dtype))
+                rot_z = so3.exp(torch.tensor([[0, 0, rpy[2]]], dtype=dtype))
+                origin[:, :, :3] = rot_x @ rot_y @ rot_z
+
+            return origin
+
+        urdf_model = urdf.URDF.from_xml_file(urdf_file)
+        robot = Robot(urdf_model.name, dtype)
+
+        for urdf_link in urdf_model.links:
+            link = Link(urdf_link.name)
+            robot._link_map[urdf_link.name] = link
+
+        for urdf_joint in urdf_model.joints:
+            origin = get_origin(urdf_joint.origin)
+            joint_type = get_joint_type(urdf_joint)
+            parent = robot.link_map[urdf_joint.parent]
+            child = robot.link_map[urdf_joint.child]
+            if joint_type == RevoluteJoint or joint_type == PrismaticJoint:
+                axis = torch.tensor(urdf_joint.axis, dtype=dtype)
+                joint = joint_type(
+                    urdf_joint.name,
+                    axis,
+                    parent=parent,
+                    child=child,
+                    origin=origin,
+                )
+            else:
+                joint = joint_type(
+                    urdf_joint.name, parent=parent, child=child, origin=origin
+                )
+            child.set_parent(joint)
+            parent.add_child(joint)
+            robot._joint_map[urdf_joint.name] = joint
+
+        for _, link in robot.link_map.items():
+            for joint in link._children:
+                if isinstance(joint, FixedJoint):
+                    subjoints: List[Joint] = joint.child.children
+                    joint.child.set_children([])
+                    for subjoint in subjoints:
+                        subjoint.set_parent(link)
+                        subjoint.set_origin(se3.compose(joint.origin, subjoint.origin))
+                        link.children.append(subjoint)
+
+        joints_to_visit: List[Joint] = []
+        root = robot.link_map[urdf_model.get_root()]
+        num_joints = 0
+        root.set_id(0)
+        robot._links.append(root)
+        robot._dof = 0
+        joints_to_visit = joints_to_visit + root.children
+
+        while joints_to_visit:
```

Comment:
Add some comments here explaining what this does. Here is what I understood from our conversation:

* This part creates IDs for all joints and links, and appends the nodes to the `_joints` and `_links` lists.
* The robot's dof only depends on non-fixed joints.
* Fixed joints are processed after non-fixed ones to make their IDs larger values.

Reply:
Added